### PR TITLE
Enable MerlinAU Script Auto-Updates and Fix Backupmon Catch

### PR DESCRIPTION
**Pull request by ExtremeFiretop**
Enable MerlinAU Script Auto-Updates and Fix Backupmon Catch

### CODE REVIEW

File: MerlinAU.sh [hunk 1]
```diff
@@ -6015,6 +6154,19 @@ Please manually update to version ${GRNct}${MinSupportedFirmwareVers}${NOct} or
...
+    if [ "$retCode" -ne 0 ]
+    then
+        Say "\n${REDct}**IMPORTANT NOTICE**:${NOct}\n"
+        Say "The firmware flash has been ${REDct}CANCELLED${NOct} due to a failed backup from BACKUPMON.\n"
+        Say "Please fix the BACKUPMON configuration, or consider uninstalling it to proceed flash.\n"
+        Say "Resolving the BACKUPMON configuration is HIGHLY recommended for safety of the upgrade.\n"
+        "$inMenuMode" && _WaitForEnterKey_ "$mainMenuReturnPromptStr"
+        _Reset_LEDs_
+        return 1
+    fi
```
Comment (ExtremeFiretop) [inline MerlinAU.sh:6168]: IMPORTANT FIX!

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

The only issue I have with this solution right now is the cron for the update check isnt dynamically following the regular Cron job (by being 15 minutes ahead)

It's 15 minutes ahead by default, but someone could set the regular Cron job for firmware updates to the same time with this solution.

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

Any reason why the 15 minutes ahead of a hard requirement? What if we simply piggy backed in the existing cron job? And just made sure the update happens before the firmware flash?

**Comment by ExtremeFiretop** [comment]
@Martinski4GitHub 

Checkout: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/365

 Over 100 lines of code less and it's basically ready, in comparsion to the 2 cron job method 15 minutes part which is 85% of the way? Maybe?

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

My apologies for the delayed response. Yesterday was a long and busy day for me and after dinner time I felt very tired so I went to bed early. 

I'll take a look at and review your PRs later this evening. Now on to my day job :>)

Take care, bud.

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

Well, my wife & I got a nasty case of cold this evening. Yesterday evening, my wife felt a little under the weather & I just felt very tired, but now I think our bodies were already giving us a warning as it was fighting the cold virus multiplying in our systems. Then today, by the end of the afternoon, we both had full-on cold symptoms. At least it's not flu, but we have coughs, sneezing & watery eyes. We already took some medications to help us go to sleep.

Give me a couple of days to recover from this cold.
Good night, bud.

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop,
> 
> Well, my wife & I got a nasty case of cold this evening. Yesterday evening, my wife felt a little under the weather & I just felt very tired, but now I think our bodies were already giving us a warning as we were fighting the cold virus multiplying in our systems. Then today, by the end of the afternoon, we both had full-on cold symptoms. At least it's not flu, but we have coughs, sneezing & watery eyes. We already took some medications to help us go to sleep.
> 
> Give me a couple of days to recover from this cold. Good night, bud.

I hope you feel better soon buddy. No rush, both PRs do the exact same thing, the question is on which method to use now.

But I'll wait for you to bounce some ideas around , I have some ideas around both methods (2 seperate Cron jobs or using the existing one)

Using the existing one is the easiest obviously since it's less code and we don't need to worry about 2 Cron jobs overlapping. But I also thought about your new function and how it can deal with the lock file even if 2 Cron jobs do call MerlinAU so maybe you'll want to utilize that with the other method of using 2 Cron jobs?

Both methods have their merits, something to think about for when you recover from best rest 🙏

**Comment by Martinski4GitHub** [comment]
> > @ExtremeFiretop,
> > Well, my wife & I got a nasty case of cold this evening. Yesterday evening, my wife felt a little under the weather & I just felt very tired, but now I think our bodies were already giving us a warning as we were fighting the cold virus multiplying in our systems. Then today, by the end of the afternoon, we both had full-on cold symptoms. At least it's not flu, but we have coughs, sneezing & watery eyes. We already took some medications to help us go to sleep.
> > Give me a couple of days to recover from this cold. Good night, bud.
> 
> I hope you feel better soon buddy. No rush, both PRs do the exact same thing, the question is on which method to use now. I created the first PR based on your comment of 2 cron jobs, and the second based on an idea to avoid the overlapping situation (using the existing cron job)
> 
> But I'll wait for you to bounce some ideas around , I have some ideas around both methods (2 seperate Cron jobs or using the existing one)
> 
> Using the existing one is the easiest obviously since it's less code and we don't need to worry about 2 Cron jobs overlapping. But I also thought about your new function and how it can deal with the lock file even if 2 Cron jobs do call MerlinAU so maybe you'll want to utilize that with the other method of using 2 Cron jobs?
> 
> Both methods have their merits, something to think about for when you recover from bed rest 🙏
>
Thanks, bud. We certainly need the rest. We're still in the midst of the cold storm, but the medicine to sleep a little better works fairly well. When I get a cold, it usually lasts 3 to 4 days at most, so we should be OK by Friday - fingers crossed.

I have not yet reviewed your PRs. I started to take a look but didn't get very far with the watery eyes and stuffy nose just making it uncomfortable to look at the screen, concentrate & think about what I'm reading. I'll try again tomorrow if I can.

Now, I'm just going to sleep early after taking the cold medicine and getting some VaporRub for the chest & nose.
Take care, bud & good night.

**Comment by ExtremeFiretop** [comment]
> 
> Now, I'm just going to sleep early after taking the cold medicine and getting some VaporRub for the chest & nose. Take care, bud & good night.

Hope your starting to feel better @Martinski4GitHub ! VapoRub and some cough medicine is the winnings combination to help with sleep. 😆 

Also breath right nose strips if your really congested! Not being able to sleep while sick and tired is one of the worst feelings.

**Comment by Martinski4GitHub** [comment]
> > Now, I'm just going to sleep early after taking the cold medicine and getting some VaporRub for the chest & nose. Take care, bud & good night.
> 
> Hope your starting to feel better @Martinski4GitHub ! VapoRub and some cough medicine is the winnings combination to help with sleep. 😆
> 
> Also breathe right nose strips if your really congested! Not being able to sleep while sick and tired is one of the worst feelings.
>
>
Thank you. My wife & I have been feeling much better since yesterday evening, and we had a really good sleep last night. Today in the afternoon, we felt fully recovered and back to "normal" & back to "business as usual." :>). In fact, we went shopping, including getting groceries and all the stuff that we'll need for the Thanksgiving celebration coming up here in the USA, and then we went out to a French Bistro ("Bistro De La Gare") for dinner. Delicious!!!!

**Comment by Martinski4GitHub** [comment]
@ExtremeFiretop,

I've reviewed both PRs and I prefer the one that sets up 2 separate cron jobs, one for each automatic update. I believe that quite a few users will likely want to have different intervals & days to check for automatic updates.

For example, in my particular case, normally my cron job is configured to run early in the morning on Saturdays & Sundays only ("**10 3 * * Sat,Sun**"). The reason is that I prefer to deal with any possible issues that might come up due to the F/W update during the weekends rather than wake up on a weekday without an internet connection because something went wrong or because I have to do a factory default reset. It happens rarely but it has happened before. This is especially important for people who work from home.

In my case, if I were not one of the developers of the script, I would prefer to have a separate cron job to check for the automatic script update every day in the early morning (i.e. "30 2 * * *") while maintaining my F/W Update checks on the weekends only.

My 2 cents.

**Comment by Martinski4GitHub** [comment]
> @Martinski4GitHub
> 
> The only issue I have with this solution right now is the cron for the update check isnt dynamically following the regular Cron job (by being 15 minutes ahead)
> 
> It's 15 minutes ahead by default, but someone could set the regular Cron job for firmware updates to the same time with this solution.
>
>
I have an idea of how to deal with synchronizing the 2 cron jobs to run 15 minutes apart based on the setup of the F/W Update cron job. Let me work on it and I'll submit a PR for your review by tomorrow evening.

**Comment by ExtremeFiretop** [comment]
> > > Now, I'm just going to sleep early after taking the cold medicine and getting some VaporRub for the chest & nose. Take care, bud & good night.
> > 
> > 
> > Hope your starting to feel better @Martinski4GitHub ! VapoRub and some cough medicine is the winnings combination to help with sleep. 😆
> > Also breathe right nose strips if your really congested! Not being able to sleep while sick and tired is one of the worst feelings.
> 
> Thank you. My wife & I have been feeling much better since yesterday evening, and we had a really good sleep last night. Today in the afternoon, we felt fully recovered and back to "normal" & back to "business as usual." :>). In fact, we went shopping, including getting groceries and all the stuff that we'll need for the Thanksgiving celebration coming up here in the USA, and then we went out to a French Bistro ("Bistro De La Gare") for dinner. Delicious!!!!

I was just heading to bed when I got this notification! Happy to hear your feeling better and out and about again! Always good to be back in business! 

Sounds like a great meal! 😋 I ordered a steak but it may a been a little too red for me unfortunately. Otherwise meal was delicious!

**Review by Martinski4GitHub — APPROVED** [review]
I like this solution better!!

**Comment by ExtremeFiretop** [comment]
> @ExtremeFiretop,
> 
> I've reviewed both PRs and I prefer the one that sets up 2 separate cron jobs, one for each automatic update. I believe that quite a few users will likely want to have different intervals & days to check for automatic updates.
> 
> For example, in my particular case, normally my cron job is configured to run early in the morning on Saturdays & Sundays only ("**10 3 * * Sat,Sun**"). The reason is that I prefer to deal with any possible issues that might come up due to the F/W update during the weekends rather than wake up on a weekday without an internet connection because something went wrong or because I have to do a factory default reset. It happens rarely but it has happened before. This is especially important for people who work from home.
> 
> In my case, if I were not one of the developers of the script, I would prefer to have a separate cron job to check for the automatic script update every day in the early morning (i.e. "30 2 * * *") while maintaining my F/W Update checks on the weekends only.
> 
> My 2 cents.

Totally understand that view point, I just figured the other solution would be simpler to implement if we piggy backed on the existing cron, and it would guarantee we don't overlap.

But as you know, I'm open to suggestions it's why I gave both as an available option really, I didn't know which you'd rather run with to be quite honest. 

Both have merits and it really depends on the individual, I am fine with either method. (As long as we find a suitable solution to the overlapping cron jobs)

**Comment by Martinski4GitHub** [comment]
> > @ExtremeFiretop,
> > I've reviewed both PRs and I prefer the one that sets up 2 separate cron jobs, one for each automatic update. I believe that quite a few users will likely want to have different intervals & days to check for automatic updates.
> > For example, in my particular case, normally my cron job is configured to run early in the morning on Saturdays & Sundays only ("**10 3 * * Sat,Sun**"). The reason is that I prefer to deal with any possible issues that might come up due to the F/W update during the weekends rather than wake up on a weekday without an internet connection because something went wrong or because I have to do a factory default reset. It happens rarely but it has happened before. This is especially important for people who work from home.
> > In my case, if I were not one of the developers of the script, I would prefer to have a separate cron job to check for the automatic script update every day in the early morning (i.e. "30 2 * * *") while maintaining my F/W Update checks on the weekends only.
> > My 2 cents.
> 
> Totally understand that view point, I just figured the other solution would be simpler to implement if we piggy backed on the existing cron, and it would guarantee we don't overlap.
> 
> But as you know, I'm open to suggestions it's why I gave both as an available option really, I didn't know which you'd rather run with to be quite honest.
> 
> Both have merits and it really depends on the individual, I am fine with either method. (As long as we find a suitable solution to the overlapping cron jobs)
>
>
OK, good. I'll merge this PR and start working on a solution to avoid any overlap of the 2 cron jobs.

**Comment by ExtremeFiretop** [comment]
> > @Martinski4GitHub
> > The only issue I have with this solution right now is the cron for the update check isnt dynamically following the regular Cron job (by being 15 minutes ahead)
> > It's 15 minutes ahead by default, but someone could set the regular Cron job for firmware updates to the same time with this solution.
> 
> I have an idea of how to deal with synchronizing the 2 cron jobs to run 15 minutes apart based on the setup of the F/W Update cron job. Let me work on it and I'll submit a PR for your review by tomorrow evening.

Take your time, I thought about your new function and we can probably utilize that. (If they do happen to fire at once they can wait for the update job first , then do the firmware update)

But your the master, take your time. I'll be around to discuss whatever your mind thinks up tomorrow 😋

**Pull request merged**